Keep maintenance mode on if it was previously enabled

jokeputs opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->
When enabling maintenance mode it will get disabled after using one of the following commands:
- `bin\magento module:uninstall`
- `bin\magento setup:backup`
- `bin\magento setup:rollback`
- `bin\magento theme:uninstall`
- `bin\magento deploy:mode:set production`

The changes in this PR will first check or maintenance mode was already enabled. If so, it will skip disabling maintenance mode after executing the command. 

### Fixed Issues
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#9918: Magento 2 automatically disables maintenance mode after certain actions

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. Enable maintenance mode: `bin\magento maintenance:enable`
2. Create a backup: `bin/magento setup:backup --media`
3. Confirm that maintenance mode is still enabled: `bin\magento maintenance:status`

The output will be for example:
```
Maintenance mode already enabled
Media backup is starting...
Media backup filename: 1506422834_filesystem_media.tgz (The archive can be uncompressed with 7-Zip on Windows systems)
Media backup path: /var/www/var/backups/1506422834_filesystem_media.tgz
[SUCCESS]: Media backup completed successfully.
Skipped disabling maintenance mode
```

jokeputs committed:
Keep maintenance mode on if it was previously enabled

antonkril commented on app/code/Magento/Theme/Console/Command/ThemeUninstallCommand.php:
```diff
@@ -375,4 +378,40 @@ private function cleanup(InputInterface $input, OutputInterface $output)
+     * @param OutputInterface $output
+     * @return void
+     */
+    private function enableMaintenanceMode(OutputInterface $output)
```
Code duplication can be avoided by extracting this behavior to a separate object

jokeputs replied:
How would you do this? Create something like a `MaintenanceModeManagement`? Or would you add this feature to `Magento\Framework\App\MaintenanceMode`? I considered that last one but I don't think `MaintenanceMode` should be responsible for knowing or it should disable maintenance mode.
Do you already have something in mind?

jokeputs committed:
Moved all maintenance logic to MaintenanceModeEnabler

jokeputs committed:
Updated tests to include the MaintenanceModeEnabler

jokeputs commented:
@antonkril, I've updated the pull request. All logic was mode to `Magento\Framework\App\Console\MaintenanceModeEnabler`.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=11052) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=11052) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: jokeputs<br/>:x: paliarush<br/>:x: viktym

jokeputs committed:
Merge branch '2.2-develop' into keep-maintenance-enabled

vkublytskyi commented on app/code/Magento/Deploy/Model/Mode.php:
```diff
@@ -83,7 +83,7 @@ class Mode
      * @param Writer $writer
      * @param Reader $reader
-     * @param MaintenanceMode $maintenanceMode
+     * @param MaintenanceModeEnabler $maintenanceMode
```
Such change is backward incompatible and not allowed for a patch releases. You should deprecate old constructor parameter and [add new one](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/#adding-a-constructor-parameter).

vkublytskyi reviewed (changes requested):
@jokeputs Changing of constructor arguments is a backward incompatible change and should be avoided. See inline comment how it may be resolved.

Also, some little code duplication is still present as for each command we need to enable and disable mode.
In general, Magento does not allow introduce new public methods in patch releases. But in this case, I suppose this is ok so I propose to add new method to 
`Magento\Framework\App\MaintenanceMode`:

```php
public function executeInMaintainanceMode(callable $task)
{
    // check current maintenance mode
    // switch on maintenance mode if not yet enabled
    call_user_func($task);
    // restore initial mode if needed
}
```

Such approach will reduce code duplication to minimum an will not lead to a lot of code changes and constructor arguments (only wrapping some part of logic in anonymous function is required)

vkublytskyi commented on lib/internal/Magento/Framework/App/Console/MaintenanceModeEnabler.php:
```diff
@@ -0,0 +1,96 @@
+     * @param OutputInterface $output
+     * @return void
+     */
+    private function enableMaintenanceMode(OutputInterface $output)
```
I made this function private as it changes internal state. So class interfaces with enable/disable methods creates temporal coupling in a code.

Class `MaintenanceModeEnabler` is preserved in a code as it would be wrong to introduce dependency on `OutputInterface` in `MaintenanceMode` (initial recommendation).

Commit:
MAGETWO-80111: Keep maintenance mode on if it was previously enabled #11052

 - revert backward incompatible changes
 - fix temporal coupling in \Magento\Framework\App\Console\MaintenanceModeEnabler
 - cover maintenance mode with unit test
 - revert changes in unit test to ensure code have same behavior after refactoring
